lib/service/discovery.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/discovery"
+	"github.com/gravitational/trace"
+)
+
+func (process *TeleportProcess) shouldInitDiscovery() bool {
+	return process.Config.Discovery.Enabled && len(process.Config.Discovery.AWSMatchers) != 0
+}
+
+func (process *TeleportProcess) initDiscovery() {
+	process.registerWithAuthServer(types.RoleDiscovery, DiscoveryIdentityEvent)
+	process.RegisterCriticalFunc("discovery.init", process.initDiscoveryService)
+}
+
+func (process *TeleportProcess) initDiscoveryService() error {
+	log := process.log.WithField(trace.Component, teleport.Component(
+		teleport.ComponentDiscovery, process.id))
+
+	conn, err := process.waitForConnector(DiscoveryIdentityEvent, log)
```

Comment:
By making a new `DiscoveryIdentityEvent`, aren't we requiring node join tokens to have a new role for the discovery service? If so, this is going to break a lot of the existing docs.

Should discovery service instead try to use the identity of the SSH service?

(Honest question - I have a similar challenge with the uploader service)

Reply:
@zmb3 The idea is to have a separate discovery service which you can run alongside other services (app, db, ssh) or separately. Reason is discovery service is a standalone component that just discovers resources and registers them in a cluster, with its own set of permissions and caches that it needs, etc. I had this among the topics for the today's product sync.

For the uploader service it seem that it can't really run as standalone and should be an internal service so the use-case is a little different.

Reply:
There's lots of places in the docs where we instruct users how to add a node to a cluster. We'll want to make sure to mention the new discovery role, otherwise people who enable discovery service in the config without ensuring their host certs have the discovery role will run into errors and discovery won't work.